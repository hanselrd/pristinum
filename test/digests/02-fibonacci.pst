func fib(n @u64) @u64:
    if n 2 <:
        return n .
    end
    return fib(n 1 -) fib(n 2 -) + .
end

n @u64 := 1 .
while n 15 <:
    print(fib(n)) .
    n = n 1 + .
end