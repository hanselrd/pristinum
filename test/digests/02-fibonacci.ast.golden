PProgram
    [ SFunction
        ( IDENT "fib" )
        [ BBind
            ( IDENT "n" ) TUint64
        ] TUint64
        [ SIf
            ( ELess
                ( EIdent
                    ( IDENT "n" )
                )
                ( EInt 2 )
            )
            [ SReturn
                ( EIdent
                    ( IDENT "n" )
                )
            ] []
        , SReturn
            ( EAdd
                ( ECall
                    ( IDENT "fib" )
                    [ ESubtract
                        ( EIdent
                            ( IDENT "n" )
                        )
                        ( EInt 1 )
                    ]
                )
                ( ECall
                    ( IDENT "fib" )
                    [ ESubtract
                        ( EIdent
                            ( IDENT "n" )
                        )
                        ( EInt 2 )
                    ]
                )
            )
        ]
    , SBind
        ( BBind
            ( IDENT "n" ) TUint64
        )
        ( EInt 1 )
    , SWhile
        ( ELess
            ( EIdent
                ( IDENT "n" )
            )
            ( EInt 15 )
        )
        [ SExpr
            ( ECall
                ( IDENT "print" )
                [ ECall
                    ( IDENT "fib" )
                    [ EIdent
                        ( IDENT "n" )
                    ]
                ]
            )
        , SAssign
            ( IDENT "n" )
            ( EAdd
                ( EIdent
                    ( IDENT "n" )
                )
                ( EInt 1 )
            )
        ]
    ]